pkg/ccl/backupccl/show.go

```diff
@@ -448,18 +450,115 @@ you must pass the 'encryption_info_dir' parameter that points to the directory o
 						len(dest)))
 			}
 		}
+
+		if _, ok := opts[backupOptCheckFiles]; ok {
+			fileSizes, err := checkBackupFiles(ctx, info,
+				p.ExecCfg().DistSQLSrv.ExternalStorageFromURI,
+				p.User())
+			if err != nil {
+				return err
+			}
+			info.fileSizes = fileSizes
+		}
 		return infoReader.showBackup(ctx, &mem, mkStore, info, p.User(), resultsCh)
 	}
 
 	return fn, infoReader.header(), nil, false, nil
 }
 
+// checkBackupFiles validates that each SST is in its expected storage location
+func checkBackupFiles(
+	ctx context.Context,
+	info backupInfo,
+	storeFactory cloud.ExternalStorageFromURIFactory,
+	user security.SQLUsername,
+) ([][]int64, error) {
+
+	checkLayer := func(layer int) ([]int64, error) {
+		// TODO (msbutler): Right now, checkLayer opens stores for each backup layer. In 22.2,
+		// once a backup chain cannot have mixed localities, only create stores for full backup
+		// and first incremental backup.
+		defaultStore, err := storeFactory(ctx, info.defaultURIs[layer], user)
+		if err != nil {
+			return nil, err
+		}
+		localityStores := make(map[string]cloud.ExternalStorage)
+
+		defer func() {
+			if err := defaultStore.Close(); err != nil {
+				log.Warningf(ctx, "close export storage failed %v", err)
+			}
+			for _, store := range localityStores {
+				if err := store.Close(); err != nil {
+					log.Warningf(ctx, "close export storage failed %v", err)
+				}
+			}
+		}()
+		// Check metadata files. Note: we do not check locality aware backup
+		// metadata files ( prefixed with `backupPartitionDescriptorPrefix`) , as
+		// they're validated in resolveBackupManifests.
+		for _, metaFile := range []string{
+			fileInfoPath,
+			metadataSSTName,
+			backupStatisticsFileName,
+			backupManifestName + backupManifestChecksumSuffix} {
+			if _, err := defaultStore.Size(ctx, metaFile); err != nil {
+				return nil, errors.Wrapf(err, "Error checking metadata file %s/%s",
+					info.defaultURIs[layer], metaFile)
+			}
+		}
+
+		for locality, uri := range info.localityInfo[layer].URIsByOriginalLocalityKV {
+			store, err := storeFactory(ctx, uri, user)
+			if err != nil {
+				return nil, err
+			}
+			localityStores[locality] = store
+		}
+
+		// Check all backup SSTs.
+		fileSizes := make([]int64, len(info.manifests[layer].Files))
+		for i, f := range info.manifests[layer].Files {
+			store := defaultStore
+			uri := info.defaultURIs[layer]
+			if _, ok := localityStores[f.LocalityKV]; ok {
```

Comment:
Are there entires in the base manifest Files slice that have non-default localities? aren't those in the other partition manifests?

Reply:
I think the default manifest contains all files, regardless of localities. Otherwise, my `SHOW BACKUP FILES` tests would fail (e.g. in `TestBackupRestoreSeperateExplicitIsDefault` around line 9713, I just verified that some files live in non default localities and were gathered from the default backup manifest). 

During Restore, it seems like the partition Manifests are only used  in `getLocalityInfo` to build a map from locality to localityURI. Notice that the `files` field from the partition manifest is never returned to the caller. Given this, I'm less clear on why we even need partition manifests

Reply:
>  I'm less clear on why we even need partition manifests

huh, me too now.